File: 二分查找/34.在排序数组中查找元素的第一个和最后一个位置.py
class Solution:
    # 通过二分查找添加偏置条件获取左右边界值
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left = self.binSearch(nums, target, True)
        right = self.binSearch(nums, target, False)
        return [left, right]

    # 添加偏置参数，在一段相同的数字中获取最左边或最右边的数字
    def binSearch(self, nums, target, leftBias):
        l, r = 0, len(nums) - 1
        i = -1  # 默认的返回的值
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] > target:
                r = mid - 1
            elif nums[mid] < target:
                l = mid + 1
            # 找到target后根据偏置条件找到边界
            else:
                i = mid
                if leftBias:
                    r = mid - 1
                else:
                    l = mid + 1
        return i

File: 二分查找/34.在排序数组中查找元素的第一个和最后一个位置.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  const left = binarySearch(nums, target, true);
  const right = binarySearch(nums, target, false);
  return [left, right];
};

function binarySearch(nums, target, leftBias) {
  let l = 0,
    r = nums.length - 1,
    mid;
  let targetIndex = -1; //! trick
  while (l <= r) {
    mid = Math.floor((l + r) / 2);
    if (target > nums[mid]) l = mid + 1;
    else if (target < nums[mid]) r = mid - 1;
    else {
      //! 在局部(连续、相同)的数组中寻找边界
      targetIndex = mid;
      if (leftBias) {
        r = mid - 1;
      } else {
        l = mid + 1;
      }
    }
  }
  return targetIndex;
}
